<a href="https://colab.research.google.com/github/PuputHanifah/PuputHanifah/blob/main/Latihan%202%20-%20Metode%20Numerik%20(Fixed%20Point)%20-%20Puput%20Hanifah%20Nuranisa%20-%2011210940000017.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Latihan 2 Metode Numerik (Fixed-Point)**

**Nama : Syarifah Syila Alkadrie**

**NIM : 11210940000019**


#Soal A

In [ ]:
import math

def f(x):
  a1 = x*x*x + 4*x*x -10
  return a1

def g(x):
  a2 = x- x*x*x - 4*x*x + 10
  return a2

def IterasiFixedPoint(x0, epsilon, N):
  print('\nBerikut adalah Iterasi Fixed Point dari Persamaan diatas:')
  langkah = 1
  flag = 1
  kondisi = True
  while kondisi:
    x1 = g(x0)
    print('Iteration-%d, x1 = %0.6f dan f(x1) = %0.6f' % (langkah, x1, f(x1)))
    x0 = x1

    langkah = langkah + 1
    if langkah > N:
      flag = 0
      break

    kondisi = abs(f(x1)) > epsilon

  if flag == 1:
    print('\nSolusi atau nilai akar yang didapatkan adalah: %0.8f' % x1)
  else:
    print('\nIterasi Fixed Point diatas bersifat Divergen.')

x0 = float(input('Tebakan Awal (x0): '))
epsilon = float(input('Toleransi Error: '))
N = int(input('Iterasi Maksimal: '))

IterasiFixedPoint(x0, epsilon, N)
                  

Tebakan Awal (x0): 1.5
Toleransi Error: 0.00001
Iterasi Maksimal: 5

Berikut adalah Iterasi Fixed Point dari Persamaan diatas:
Iteration-1, x1 = -0.875000 dan f(x1) = -7.607422
Iteration-2, x1 = 6.732422 dan f(x1) = 476.452434
Iteration-3, x1 = -469.720012 dan f(x1) = -102755024.907397
Iteration-4, x1 = 102754555.187385 dan f(x1) = 1084933870531746486812672.000000
Iteration-5, x1 = -1084933870531746352594944.000000 dan f(x1) = -1277055591444378074254579861314550183250535909418315265493330570988486656.000000

Iterasi Fixed Point diatas bersifat Divergen.


**Penjelasan :**
1.   Persamaan g(x) diatas didapat berdasarkan pendefinisian fungsi g yang dapat dituliskan, yaitu x = g(x) = x -
f(x), dimana f(x) = x^3 + 4x^2 - 10 = 0, sehingga g(x) = x - x^3 - 4x^2 + 10.

2.   Berdasarkan hasil iterasi fixed point diatas, dapat disimpulkan bahwa untuk persamaan g(x) diatas pada
iterasi ke-4 dan seterusnya menghasilnya nilai yang mendekati infinity (tak hingga) sehingga mengakibatkan
iterasi fixed point nya menjadi divergen.


#Soal B


In [ ]:
import math
import numpy as np

def f(x):
  b1 = x*x*x + 4*x*x -10
  return b1

def g(x):
  b2 = np.sqrt((10/x)-4*x)
  return b2

def IterasiFixedPoint(x0, epsilon, N):
  print('\nBerikut adalah Iterasi Fixed Point dari Persamaan diatas:')
  langkah = 1
  flag = 1
  kondisi = True
  while kondisi:
    x1 = g(x0)
    print('Iteration-%d, x1 = %0.6f dan f(x1) = %0.6f' % (langkah, x1, f(x1)))
    x0 = x1

    langkah = langkah + 1

    if langkah > N:
      flag = 0
      break

    kondisi = abs(f(x1)) > epsilon

  if flag == 1:
    print('\nSolusi atau nilai akar yang didapatkan adalah: %0.8f' % x1)
  else:
    print('\nIterasi Fixed Point diatas bersifat Divergen.')

x0 = float(input('Tebakan Awal (x0): '))
epsilon = float(input('Toleransi Error: '))
N = int(input('Iterasi Maksimal: '))

IterasiFixedPoint(x0, epsilon, N)


Tebakan Awal (x0): 1.5
Toleransi Error: 0.00001
Iterasi Maksimal: 5

Berikut adalah Iterasi Fixed Point dari Persamaan diatas:
Iteration-1, x1 = 0.816497 dan f(x1) = -6.789002
Iteration-2, x1 = 2.996909 dan f(x1) = 52.842473
Iteration-3, x1 = nan dan f(x1) = nan

Solusi atau nilai akar yang didapatkan adalah: nan


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:9: RuntimeWarning: invalid value encountered in sqrt
  if __name__ == '__main__':


**Penjelasan :**


1.   Persamaan g(x) diatas didapatkan dari hasil manipulasi aljabar sederhana terhadap f(x) = x^3 + 4x^2 - 10 = 0,
yaitu:

  4x^2 = 10 - x^3

  x^3 = 4x^2 - 10

  x^2 = 4x - (10/x)

  x = +- [4x - (10/x)]^(0.5)

  Dalam kasus ini diambil nilai x yang negatif, yaitu: x = [(10/x) - 4x]^(1/2) = g(x)

2.   Berdasarkan hasil iterasi diatas, dapat disimpulkan bahwa untuk persamaan g(x) diatas berhenti pada iterasi
ke-3 karena pada iterasi ke-3 menghasilkan nilai yang imajiner (i) sehingga mengakibatkan iterasi fixed point
diatas menjadi divergen.




#Soal C

In [ ]:
import math

def f(x):
  c1 = x*x*x + 4*x*x -10
  return c1

def g(x):
  c2 = (0.5)*(10-x**3)**(0.5)
  return c2

def IterasiFixedPoint(x0, epsilon, N):
  print('\nBerikut adalah Iterasi Fixed Point dari Persamaan diatas:')
  langkah = 1
  flag = 1
  kondisi = True
  while kondisi:
    x1 = g(x0)
    print('Iteration-%d, x1 = %0.6f dan f(x1) = %0.6f' % (langkah, x1, f(x1)))
    x0 = x1

    langkah = langkah + 1

    if langkah > N:
      flag = 0
      break

    kondisi = abs(f(x1)) > epsilon

  if flag == 1:
    print('\nSolusi atau nilai akar yang didapatkan adalah: %0.8f' % x1)
  else:
    print('\nIterasi Fixed Point diatas bersifat Divergen.')

x0 = float(input('Tebakan Awal (x0): '))
epsilon = float(input('Toleransi Error: '))
N = int(input('Iterasi Maksimal: '))

IterasiFixedPoint(x0, epsilon, N)

Tebakan Awal (x0): 1.5
Toleransi Error: 0.00001
Iterasi Maksimal: 25

Berikut adalah Iterasi Fixed Point dari Persamaan diatas:
Iteration-1, x1 = 1.286954 dan f(x1) = -1.243483
Iteration-2, x1 = 1.402541 dan f(x1) = 0.627450
Iteration-3, x1 = 1.345458 dan f(x1) = -0.323340
Iteration-4, x1 = 1.375170 dan f(x1) = 0.164948
Iteration-5, x1 = 1.360094 dan f(x1) = -0.084596
Iteration-6, x1 = 1.367847 dan f(x1) = 0.043270
Iteration-7, x1 = 1.363887 dan f(x1) = -0.022163
Iteration-8, x1 = 1.365917 dan f(x1) = 0.011344
Iteration-9, x1 = 1.364878 dan f(x1) = -0.005808
Iteration-10, x1 = 1.365410 dan f(x1) = 0.002973
Iteration-11, x1 = 1.365138 dan f(x1) = -0.001522
Iteration-12, x1 = 1.365277 dan f(x1) = 0.000779
Iteration-13, x1 = 1.365206 dan f(x1) = -0.000399
Iteration-14, x1 = 1.365242 dan f(x1) = 0.000204
Iteration-15, x1 = 1.365224 dan f(x1) = -0.000105
Iteration-16, x1 = 1.365233 dan f(x1) = 0.000054
Iteration-17, x1 = 1.365228 dan f(x1) = -0.000027
Iteration-18, x1 = 1.365231 dan f(x1) =

**Penejelasan :**


1.   Persamaan g(x) diatas didapatkan dari hasil manipulasi aljabar sederhana terhadap f(x) = x^3 + 4x^2 - 10 = 0,
yaitu:

  4x^2 = 10 - x^3

  x^2 = (1/4).(10 - x^3)

  x = +-(1/2).(10 - x^3)

  Dalam kasus ini diambil nilai x yang positif, yaitu: x = (1/2) (10 - x^3) = g(x)
2.   Berdasarkan hasil iterasi diatas, dapat disimpulkan bahwa untuk persamaan g(x) diatas dikatakan konvergen
karena menghasilkan solusi akar dari persamaan f(x), yaitu 1.36522958.



#Soal D

In [ ]:
import math

def f(x):
  d1 = x*x*x + 4*x*x -10
  return d1

def g(x):
  d2 = (10/(4+x))**(0.5)
  return d2

def IterasiFixedPoint(x0, epsilon, N):
  print('\nBerikut adalah Iterasi Fixed Point dari Persamaan diatas:')
  langkah = 1
  flag = 1
  kondisi = True
  while kondisi:
    x1 = g(x0)
    print('Iteration-%d, x1 = %0.6f dan f(x1) = %0.6f' % (langkah, x1, f(x1)))
    x0 = x1

    langkah = langkah + 1

    if langkah > N:
      flag = 0
      break

    kondisi = abs(f(x1)) > epsilon

  if flag == 1:
    print('\nSolusi atau nilai akar yang didapatkan adalah: %0.8f' % x1)
  else:
    print('\nIterasi Fixed Point diatas bersifat Divergen.')

x0 = float(input('Tebakan Awal (x0): '))
epsilon = float(input('Toleransi Error: '))
N = int(input('Iterasi Maksimal: '))

IterasiFixedPoint(x0, epsilon, N)

Tebakan Awal (x0): 1.5
Toleransi Error: 0.00001
Iterasi Maksimal: 15

Berikut adalah Iterasi Fixed Point dari Persamaan diatas:
Iteration-1, x1 = 1.348400 dan f(x1) = -0.275637
Iteration-2, x1 = 1.367376 dan f(x1) = 0.035481
Iteration-3, x1 = 1.364957 dan f(x1) = -0.004508
Iteration-4, x1 = 1.365265 dan f(x1) = 0.000574
Iteration-5, x1 = 1.365226 dan f(x1) = -0.000073
Iteration-6, x1 = 1.365231 dan f(x1) = 0.000009

Solusi atau nilai akar yang didapatkan adalah: 1.36523058


**Penjelasan :**


1.   Persamaan g(x) diatas didapatkan dari hasil manipulasi aljabar sederhana terhadap f(x) = x^3 + 4x^2 - 10 = 0,
yaitu:

  4x^2 = 10 - x^3

  4x^2 + x^3 = 10

  x^2 (4 + x) = 10

  x = +-[10/(4 + x)]^(1/2)

  Dalam kasus ini diambil nilai x yang positif, yaitu: x = [10/(4 + x)]^(1/2) = g(x
2.   Berdasarkan hasil iterasi diatas, dapat disimpulkan bahwa untuk persamaan g(x) diatas dikatakan konvergen
karena menghasilkan solusi akar dari persamaan f(x), yaitu 1.36523058.



#Soal E

In [ ]:
import math

def f(x):
  e1 = x*x*x + 4*x*x -10
  return e1

def g(x):
  e2 = x-((x**3+4*x**2-10)/(3*x**2+8*x))
  return e2

def IterasiFixedPoint(x0, epsilon, N):
  print('\nBerikut adalah Iterasi Fixed Point dari Persamaan diatas:')
  langkah = 1
  flag = 1
  kondisi = True
  while kondisi:
    x1 = g(x0)
    print('Iteration-%d, x1 = %0.6f dan f(x1) = %0.6f' % (langkah, x1, f(x1)))
    x0 = x1

    langkah = langkah + 1

    if langkah > N:
      flag = 0
      break

    kondisi = abs(f(x1)) > epsilon

  if flag == 1:
    print('\nSolusi atau nilai akar yang didapatkan adalah: %0.8f' % x1)
  else:
    print('\nIterasi Fixed Point diatas bersifat Divergen.')

x0 = float(input('Tebakan Awal (x0): '))
epsilon = float(input('Toleransi Error: '))
N = int(input('Iterasi Maksimal: '))

IterasiFixedPoint(x0, epsilon, N)


Tebakan Awal (x0): 1.5
Toleransi Error: 0.00001
Iterasi Maksimal: 5

Berikut adalah Iterasi Fixed Point dari Persamaan diatas:
Iteration-1, x1 = 1.373333 dan f(x1) = 0.134345
Iteration-2, x1 = 1.365262 dan f(x1) = 0.000528
Iteration-3, x1 = 1.365230 dan f(x1) = 0.000000

Solusi atau nilai akar yang didapatkan adalah: 1.36523001


**Penjelasan :**


1.   Persamaan g(x) diatas didapatkan berdasarkan pendefinisian fungsi g dan dari hasil manipulasi aljabar yaitu
turunan dari f(x) yang dapat dituliskan: x = x - [f(x)/f'(x)] = x - [(x^3 + 4x^2 - 10) / (3x^2 + 8x)] = g(x
2.   Berdasarkan hasil iterasi diatas, dapat disimpulkan bahwa untuk persamaan g(x) diatas dikatakan konvergen
karena manghasilkan solusi akar dari persamaan f(x), yaitu 1.36523001.



#**Kesimpulan**
Dari kelima jenis persamaan g(x) diatas, terdapat beberapa hal yang
dapat disimpulkan, yaitu:


1. Iterasi fixed point dari persamaan g(x) = x - x^3 - 4x^2 + 10 dan g(x) = [(10/x) - 4x]^(1/2) menghasilkan iterasi
fixed point yang divergen.
2. Iterasi fixed point dari persamaan g(x) = (1/2) (10 - x^3), g(x) = [10 / (4 + x)]^(1/2), dan g(x) = x - [(x^3 + 4x^2 - 10)
/ (3x^2 + 8x)] menghasilkan iterasi fixed point yang konvergen karena menghasilkan solusi akar dari persamaan
f(x), yaitu 1.36523001. Dari ketiga persamaan g(x) tersebut, perbedaan hanya terletak pada jumlah iterasi yang
dibutuhkan untuk mendapatkan solusi akar dari persamaan f(x).

